File: bst/javascript/LinkedListQueue.js
class Node {
  constructor(item, nextItem) {
    this.item = item
    this.next = nextItem
  }
}

class LinkedListQueue {
  constructor() {
    this.first = null
    this.last = null
  }

  isEmpty() {
    return this.first === null
  }

  enqueue(item) {
    let oldLast = this.last
    this.last = new Node(item, null)
    if(this.isEmpty()) {
      this.first = this.last
    } else {
      oldLast.next = this.last
    }
  }

  dequeue() {
    if(this.isEmpty()) return null
    let item = this.first.item
    this.first = this.first.next
    if(this.isEmpty()) this.last = null
    return item
  }
}

File: queue/java/LinkedListQueue.java
public class LinkedListQueue<Item> {
  private Node first;
  private Node last;

  private class Node {
    Item item;
    Node next;

    Node(Item item, Node next) {
      this.item = item;
      this.next = next;
    }
  }

  public boolean isEmpty() {
    return first == null;
  }

  public void enqueue(Item item) {
    Node oldLast = last;
    last = new Node(item, null);

    if(isEmpty()) {
      first = last;
    } else {
      oldLast.next = last;
    }
  }

  public Item dequeue() {
    if(isEmpty()) return null;
    Item item = first.item;
    first = first.next;
    if(isEmpty()) last = null;
    return item;
  }
}
